hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/merge/SparkRecordMergingUtils.java

```diff
@@ -0,0 +1,189 @@
+    return FIELD_ID_TO_FIELD_MAPPING_CACHE.computeIfAbsent(avroSchema, schema -> {
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<Integer, StructField> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(fieldId, field);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * @param avroSchema Avro schema.
+   * @return The field name to ID mapping.
+   */
+  public static Map<String, Integer> getCachedFieldNameToIdMapping(Schema avroSchema) {
+    return FIELD_NAME_TO_ID_MAPPING_CACHE.computeIfAbsent(avroSchema, schema -> {
+      StructType structType = HoodieInternalRowUtils.getCachedSchema(schema);
+      Map<String, Integer> schemaFieldIdMapping = new HashMap<>();
+      int fieldId = 0;
+
+      for (StructField field : structType.fields()) {
+        schemaFieldIdMapping.put(field.name(), fieldId);
+        fieldId++;
+      }
+
+      return schemaFieldIdMapping;
+    });
+  }
+
+  /**
+   * Merges the two schemas so the merged schema contains all the fields from the two schemas,
+   * with the same ordering of fields, and the fields of the old schema comes first.
+   *
+   * @param oldSchema Old schema.
+   * @param newSchema New schema.
+   * @return The ID to {@link StructField} instance mapping of the merged schema, and the
+   * {@link StructType} and Avro schema of the merged schema.
+   */
+  public static Pair<Map<Integer, StructField>, Pair<StructType, Schema>> getCachedMergedSchema(Schema oldSchema,
+                                                                                                Schema newSchema,
+                                                                                                Schema readerSchema) {
+    return MERGED_SCHEMA_CACHE.computeIfAbsent(
+        Pair.of(Pair.of(oldSchema, newSchema), readerSchema), schemaPair -> {
+          Schema schema1 = schemaPair.getLeft().getLeft();
+          Schema schema2 = schemaPair.getLeft().getRight();
+          Schema refSchema = schemaPair.getRight();
+          Map<String, Integer> nameToIdMapping1 = getCachedFieldNameToIdMapping(schema1);
+          Map<String, Integer> nameToIdMapping2 = getCachedFieldNameToIdMapping(schema2);
+          Map<Integer, StructField> refFieldIdToFieldMapping = getCachedFieldIdToFieldMapping(refSchema);
+          Set<String> fieldNameSet = new HashSet<>();
+          fieldNameSet.addAll(nameToIdMapping1.keySet());
+          fieldNameSet.addAll(nameToIdMapping2.keySet());
+          int fieldId = 0;
+          Map<Integer, StructField> mergedMapping = new HashMap<>();
+          List<StructField> mergedFieldList = new ArrayList<>();
+          for (int i = 0; i < refFieldIdToFieldMapping.size(); i++) {
+            StructField field = refFieldIdToFieldMapping.get(i);
```

Comment:
`refFieldIdToFieldMapping` is a hash map, it does not guarantee the field sequence as what user decpares.

Reply:
As discussed, it guarantees the order of fields as the logic iterates through the field positions.  I'll add docs to explain this.